pyo3-ffi/src/lib.rs

```diff
@@ -261,7 +261,7 @@
 macro_rules! opaque_struct {
     ($name:ident) => {
         #[repr(C)]
-        pub struct $name([u8; 0]);
+        pub struct $name([u8; 0], ::std::marker::PhantomData<$crate::PyObject>);
```

Comment:
EDIT ah I see this is necessary for stable, to make the opaque structs `!Send`?

Reply:
All it does is say "this opaque struct (can) contains (pointers to) Python object(s)". It's not really needed, but in the abstract it seems like the right thing to do.

> It may be better to add separately beneath each opaque_struct usage.

`pyo3-ffi` has no idea of what `Ungil` is, because it can't depend on `pyo3`. Manual `!Ungil` impls would have to happen in `pyo3` itself, which would be error prone.

> EDIT ah I see this is necessary for stable, to make the opaque structs !Send?

That was actually not my intent -  they only exist as pointees and raw pointers are already `!Send`.

Without this phantomdata and on nightly, `Ungil` is automatically derived for things like `PyLongObject`. It means that you could create a `Py` like wrapper around `PyLongObject` and automatically have it marked `Ungil`.

That's why `impl !Ungil for ffi::PyObject` exists - without that, `Py<PyAny>` is automatically marked `Ungil`, which just seems wrong to me.

Reply:
Right, I see. I think that the ffi types should not be concerned with the `!Ungil` construct. I think that the implementations we need to implement are:
- `impl !Ungil for pyo3::PyAny`
- `impl !Ungil for pyo3::PyCell`
- `impl !Ungil for pyo3::PyRef`

`Py<PyAny>` should be `Ungil` though, as it's not tied to the GIL lifetime? `&PyAny` should be `!Ungil`.

Reply:
> I think that the ffi types should not be concerned with the !Ungil construct

This feels like "should raw pointers be `!Send` and `!Sync`?".

I think marking (pointers to) ffi types as `!Ungil` is the correct and conservative thing to do here. It leads to all those types you mentioned being `!Ungil` and also requires manual impls for things like `PyErr` and `Py<T>`. I definitely think this is the correct behaviour.


To clarify, these are the current manual implementations:

![afbeelding](https://user-images.githubusercontent.com/59372212/154143157-d1d6ca5c-e1c5-4ff9-a7d2-0609ca9071de.png)

Reply:
Ok, I take your point that the ffi types should implement `!Ungil`.

I still don't much like achieving that by adding it to this opaque struct, for a few reasons:
 - Many FFI types don't use `opaque_struct!`, especially on the unlimited API.
 - Similarly, I'm not sure that all of the FFI types which use `opaque_struct!` need to be `!Ungil`.
 - If `extern type` ever stabilises (which `opaque_struct!` is currently a placeholder for), then we'll lose all the `!Ungil` implementations if we were to swap over.

I acknowledge it's a bit painful, however I do really think we'll get a better result if we have a file with manual implementations of `!Ungil` for all the relevant FFI types.

In addition, I still think it would be a nice UX improvement to add `impl !Ungil` for `pyo3::types::PyAny`.